### PR TITLE
Up Reporter test coverage

### PR DESCRIPTION
**Pull request by c0c0n3**
### Proposed changes

This PR swaps out the QuantumLeap Docker image used in the reporter tests and replaces it with an embedded Flask server. This way `pytest` is able to collect coverage data about the `reporter` module which was previously not accessible since we were running QuantumLeap inside Docker. So the reported coverage for the `reporter` was nought whereas now it is properly reported. (Did I use report one too many times?)

### Types of changes

What types of changes does your code introduce to the project: _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

[cla]: https://fiware.github.io/contribution-requirements/individual-cla.pdf
    "FIWARE Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

**Comment by chicco785** [comment]
@c0c0n3 test are not passing, the first one seems related to the fact that crate is not ready, maybe some more time before starting the test may do the job

**Comment by amotl** [comment]
I love that improvement.

> test are not passing, the first one seems related to the fact that crate is not ready, maybe some more time before starting the test may do the job

Give me a moment to come back with a solution how to improve the situation without having to fiddle with `time.sleep()`, which will always be annoying due to introducing test flakeyness depending on CI load.

**Comment by amotl** [comment]
Hi again,

hoping that you will like it, I prepared a little walkthrough. Note that I quickly shuffled that together, so please let me know if something might be amiss.

With kind regards,
Andreas.

### Introduction
A few years ago, I would have recommended to use [`lovely.testlayers`](https://github.com/lovelysystems/lovely.testlayers) [1], which was about to run a service and then wait for it to respond on a TCP port before actually kicking off the test suite.

However, there is the excellent [`lovely-pytest-docker`](https://github.com/lovelysystems/lovely-pytest-docker) package [2] by @dobe, @burnes, @jukart and @mikethebeer (thanks!) these days, which is essentially doing the same thing, but perfectly fits into the scenario of running services within Docker containers. Specifically, it plays well with Docker Compose.

### Recipe
Here is a recipe how that might be made working for you.

#### 1. Install `lovely-pytest-docker`
Put `lovely-pytest-docker` into `requirements-test.txt` or similar.

#### 2. Adjust `conftest.py`
Put that snippet into pytest's `conftest.py` in order to define some fixtures.
```python
from crate import client
from pathlib import Path

@pytest.fixture(scope="session")
def docker_compose_files(pytestconfig):
    """
    Override this fixture in order to specify a custom location to your `docker-compose.yml`.
    """
    return Path(__file__).parent / "tests" / "docker-compose.yml"

@pytest.fixture(scope="session")
def cratedb(docker_services):
    """
    Spin up CrateDB within a Docker container by referencing
    a service name within your `docker-compose.yml`.
    """
    docker_services.start("cratedb")
    public_port = docker_services.wait_for_service("cratedb", 4200)
    url = "{docker_services.docker_ip}:{public_port}".format(**locals())
    return url

@pytest.fixture(scope="session")
def cratedb_connection(request):
    """
    Provide a CrateDB client connection to be able to execute queries.
    """
    cratedb_url = request.getfixturevalue("cratedb")
    connection = client.connect(cratedb_url)
    yield connection
    connection.close()
```

#### 3. Apply it
Use the `cratedb` or `cratedb_connection` fixture within your test cases in order to connect to CrateDB.

### Notes
While this might not be an issue on CI, the keepalive option added to `pytest` is very handy when running the tests on your workstation because it will not tear down the Docker containers on each invocation cycle. Because the test fixture layer is now responsible for the lifecycle management of the Docker containers, this option is important in order to not sacrifice your local workflow. Use it like `pytest --keepalive ...`.

**Comment by amotl** [comment]
`lovely-pytest-docker` is also capable of using a custom "is-service-up?" indicator by providing a simple probing function like `check_service_status` outlined within this example.

```python
def check_service_status(docker_ip, public_port):
    """
    Custom probing function for `wait_for_service()`.
    Waits until the service responds to HTTP requests and signals it is ready to go.
    """

    # Define URL to status endpoint.
    url = "http://{}:{}/status.json".format(docker_ip, public_port)

    # Signal "not ready" on any connection errors.
    try:
        response = requests.get(url)
    except requests.exceptions.ConnectionError:
        return False
    if not response.ok:
        return False

    # Signal readiness based on HTTP response from status endpoint.
    data = response.json()
    if "bootstrap" in data and data["bootstrap"].get("state") == "OK":
        return True

    return False
```

This can be used like:
```python
public_port = docker_services.wait_for_service("foobar-service", 1234, check_server=check_service_status)
```

**Comment by chicco785** [comment]
@amotl sounds very promising! @c0c0n3 what do you think? actually, ability to start and stop services, could be good also to properly test health checks

**Comment by chicco785** [comment]
@c0c0n3 i see the build now is ok, are you going to use the crate readiness checker and remove sleep from the bash in this PR or in a separate one?

**Comment by c0c0n3** [comment]
@amotl excellent idea, you always chip in with great suggestions! btw, if you ever want to join the coding party, give us a shout we'll gladly add you to the committer list.

So it turns out there's two issues w/ the [failed build (215954452)][travis.fail] from earlier:

1. Reporter tests. The daemon thread we fork off to run Flask needs to be sync'ed up w/ the Pytest thread that spawns it so tests only start **after** Flask has prepped the WSGI app. This is my cockup, which ddbe4ee fixed.
2. Backward compatibility tests. The call to `docker inspect` [failed][travis.docker-fail] to pick up the host address so the data load failed too and so did the test. It looks like it might have to do w/ perms on Travis, I'm looking into it.

Now back to @amotl suggestion to wait for Docker services to be up before testing. Like I said, this is a good idea in my opinion too---I think we all agree using arbitrary sleep times for that purpose is nasty and in general a non-solution :-)
I checked out `lovely-pytest-docker` and that's one hell of a lovely lib---pardon my silliness, I couldn't resist the pun. I opened #447 to implement your suggestion. Let's keep on talking about this over there.

[travis.fail]: https://travis-ci.com/github/orchestracities/ngsi-timeseries-api/builds/215954452
[travis.docker-fail]: https://travis-ci.com/github/orchestracities/ngsi-timeseries-api/jobs/480071813#L34183

**Comment by c0c0n3** [comment]
@chicco785 

> are you going to use the crate readiness checker and remove sleep from the bash in this PR or in a separate one?

I've opened #447  for that. Let's merge this PR so long if that's okay w/ you?

**Comment by chicco785** [comment]
> @chicco785
> 
> > are you going to use the crate readiness checker and remove sleep from the bash in this PR or in a separate one?
> 
> I've opened #447 for that. Let's merge this PR so long if that's okay w/ you?

totally ok, it was not just clear to me if the PR was ready or not :)

File: src/reporter/tests/embedded_server.py
Comment (chicco785) [inline src/reporter/tests/embedded_server.py:64]: a library to keep into mind also for other code parts probably

**Review by chicco785 — APPROVED** [review]
LGTM

File: src/reporter/tests/embedded_server.py
Comment (c0c0n3) [reply src/reporter/tests/embedded_server.py:64]: yep!

**Comment by c0c0n3** [comment]
> totally ok, it was not just clear to me if the PR was ready or not :)

yep, I opened a separate issue and then forgot to comment here! old age, don't judge me :-)

**Pull request merged**
